### PR TITLE
001 new delivery employee

### PR DESCRIPTION
**Pull request by TaylorJCrawford**
Post Request For New Delivery Employee.

### CODE REVIEW

File: src/main/java/org/kainos/ea/cli/EmployeeRequest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package org.kainos.ea.cli;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class EmployeeRequest {
+    private String f_name;
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/cli/EmployeeRequest.java]: These aren't the standard Java format for variables

File: src/main/java/org/kainos/ea/db/EmployeeDao.java [hunk 2]
```diff
@@ -38,4 +37,60 @@ public List<Employee> getAllEmployees() throws SQLException {
...
+
+        ResultSet rs = st.getGeneratedKeys();
+
+        if (rs.next()) {
+
+            System.out.println("Adding New Employee Delivery.");
+
+            // Add New Create ID To Employee_delivery Table
+            int newlyCreatedIDEmployeeTable = rs.getInt(1);
+            rs = null;
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/db/EmployeeDao.java]: Should this be reset to `null`, or would it be good to use a second `ResultSet`?

File: src/main/java/org/kainos/ea/cli/EmployeeRequest.java (hunk 1 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/cli/EmployeeRequest.java]: That should be all fixed now!

File: src/main/java/org/kainos/ea/db/EmployeeDao.java (hunk 2 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/db/EmployeeDao.java]: Fixed Now.

File: src/main/java/org/kainos/ea/api/EmployeeService.java [hunk 3]
```diff
@@ -9,9 +13,24 @@
...
+        try {
+            String isEmployeeRequestValid = employeeRequestValidator.isEmployeeValid(employeeRequest);
+
+            if (isEmployeeRequestValid != null) {
+                throw new EmployeeRequestIsNotValid(isEmployeeRequestValid);
+            }
+            return employeeDao.createNewEmployee(employeeRequest);
+        } catch (DeliveryEmployeeCouldNotBeCreatedException e) {
+            throw new EmployeeRequestIsNotValid(e.getMessage());
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/EmployeeService.java]: the DeliveryEmployeeCouldNotBeCreatedException is being thrown by the dao if connection to db fails so should return a 500 from controller however you're throwing a EmployeeRequestIsNotValid which is returning a 400 from controller

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/api/EmployeeService.java (hunk 3 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/api/EmployeeService.java]: Should be fixed now! If any SQL exceptions happen, we create a new sqlexception. Which should be now return a 500.

File: src/main/java/org/kainos/ea/api/EmployeeService.java [hunk 4]
```diff
@@ -41,4 +48,18 @@ public List<Employee> getAllEmployee() throws FailedToGetEmployeeException {
     }
 
 
+    public int createNewEmployee(EmployeeRequest employeeRequest) throws EmployeeRequestIsNotValid, SQLException {
+
+        try {
+            String isEmployeeRequestValid = employeeRequestValidator.isEmployeeValid(employeeRequest);
+            System.out.println("--------------------- TAYLOR HERE");
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/EmployeeService.java]: remove debug lines

File: src/main/java/org/kainos/ea/core/EmployeeRequestValidator.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+public class EmployeeRequestValidator {
+
+
+    public String isEmployeeValid (EmployeeRequest employeeRequest) {
+
+        System.out.println(employeeRequest.getfName());
+        System.out.println(employeeRequest.getlName());
+        System.out.println(employeeRequest.getNiNum());
+        System.out.println(employeeRequest.getSalary());
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/core/EmployeeRequestValidator.java]: remove these

File: src/main/java/org/kainos/ea/db/EmployeeDao.java [hunk 6]
```diff
@@ -37,14 +38,75 @@ public List<Employee> getAllEmployees() throws SQLException {
...
+                st.setInt(1, newlyCreatedIDEmployeeTable);
+                st.executeUpdate();
+                throw new DeliveryEmployeeCouldNotBeCreatedException();
+            }
+
+            // Error - Could not add employee to table.
+            throw new DeliveryEmployeeCouldNotBeCreatedException();
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new SQLException();
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/db/EmployeeDao.java]: you're throwing the same exception that your catching, did you mean to throw DeliveryEmployeeCouldNotBeCreatedException?

File: src/main/java/org/kainos/ea/resources/EmployeeController.java [hunk 7]
```diff
@@ -42,6 +46,23 @@ public Response getAllEmployeesDeliveryEmployees () {
...
     }
+
+    @POST
+    @Path("/employee/delivery")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response createNewEmployee (EmployeeRequest employeeRequest) {
+
+        try {
+            return Response.ok().entity(employeeService.createNewEmployee(employeeRequest)).build();
+        } catch (DeliveryEmployeeCouldNotBeCreatedException | EmployeeRequestIsNotValid e) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/EmployeeController.java]: DeliveryEmployeeCouldNotBeCreatedException should be returning a 500

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/resources/EmployeeController.java (hunk 7 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/resources/EmployeeController.java]: opps - sorted now???

File: src/main/java/org/kainos/ea/db/EmployeeDao.java (hunk 6 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/db/EmployeeDao.java]: nope!

File: src/main/java/org/kainos/ea/core/EmployeeRequestValidator.java (hunk 5 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/core/EmployeeRequestValidator.java]: done!

File: src/main/java/org/kainos/ea/api/EmployeeService.java (hunk 4 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/api/EmployeeService.java]: should be gone?

**Review by shaunganley — APPROVED** [review]

**Pull request merged**